pyomo/core/base/expression.py

```diff
@@ -394,6 +394,10 @@ def construct(self, data=None):
                 % (self.name, str(data))
             )
 
+        if self._anonymous_sets is not None:
+            for _set in self._anonymous_sets:
+                _set.construct()
```

Comment:
This was a concern for me, too.  At the time, I was worried about making that change (that is, requiring all components to also call construct() on their base class - in part because a lot of what I would like to move to the base class is a "prefilter" (e.g., the oft-repeated debug logging, the logic to skip-if-constructed, and now this).  But the more I think about it, the more I am inclined to propose that we implement `construct()` in the base class and have it call a `_construct_impl()` that is overridden on the derived classes.  This would reduce a lot of repeated code and shouldn't impact performance significantly.

This would be a somewhat large change.  Would you object to doing it as a separate PR?

Reply:
No objections to making that a separate PR. I'll open an issue for it.

Reply:
Tracking this in issue #3136